doc/designs/fido2.md

```diff
@@ -0,0 +1,239 @@
+the permission "System: Manage User FIDO2 credentials) can register
+credentials for another user.
+
+During the registration process, it is possible to specify
+- a COSE type: `es256`, `rs256` or `eddsa`
+- request user verification:
+the authentication will force to execute the user verification check even if
+the FIDO2 settings do not set this flag. If credentials are registered without
+the flag, the global FIDO2 settings apply.
+
+### Authentication
+
+#### Console or desktop authentication
+
+The user has a FIDO2 key in his possession, that was already registered to IPA,
+and has physical access to a machine enrolled in IPA.
+
+At Gnome login, he types his username and inserts the FIDO2 key.
+
+At console login, he types his username and inserts the FIDO2 key.
+
+SSSD validates the credentials and checks that the FIDO2 device allows
+authentication.
+
+#### Su access
+
+The user fido2user has a FIDO2 key in his possession, that was already
+registered to IPA, and has physical access to a machine enrolled in IPA. He
+is already logged into the machine as a different user and wants to perform
+su to authenticate as fido2user.
+
+Inside a terminal, he enters the `su - fido2user` command and inserts his
+FIDO2 key.
+
+SSSD validates the credentials and checks that the FIDO2 device allows
+authentication.
+
+
+## Design
+
+### Configuration of the FIDO2 settings
+
+A new LDAP entry stores the FIDO2 configuration and needs a new objectclass
+and a new attributetype:
+```
+dn: cn=fido2config,$BASEDN
+objectclass: top
+objectclass: nsContainer
+objectclass: ipafido2configObject
+cn: fido2config
+ipaRequireUserVerif: default
+```
+
+The object class allows a single attribute, request user verification,
+which is mandatory, single valued, and stores a string (on, off, default).
+The LDAP entry is added when IPA server is installed or when the server is
+upgraded to a version supporting FIDO2.
+
+### Storage of the FIDO2 key mapping
+
+The FIDO2 key mapping is stored directly in the user entry. It needs a
+new auxiliary objectclass and a new attributetype.
+
+Note: a first proposal intended to store the value in the ipasshpubkey
+attribute, but this attribute has a special handling (a new fingerprint is
+calculated for each public key and added into the attribute sshpubkeyfp)
+which makes it unsuitable for storing values that are not keys.
+
+The attribute is multi valued, optional.
+
+```
+dn: uid=idmuser,cn=users,cn=accounts,dc=ipa,dc=test
+uid: idmuser
+...
+objectClass: top
+objectClass: person
+...
+objectClass: ipafido2user
+ipafido2key: FIDO2:9S87qLk8/RxYJ3skwwYduomAM+/HDtz41N0+w/vRL6aGKJkLMsg+2OhO0E8pK5DuO1KmdK61K8PmH7jiYuOqbg==,9YE1s/f7J47h2A/DXCVFWulqoBXFzCSxcbGEBadkpSUFjwUudhPLnPUTv2qNamakXJgRYCZQ7vpS/t5zXMLnkw==
+```
+
+The FIDO2 key mapping has the format `FIDO2:credentialid,pubkey`. credential
+ID and public key are obtained during the registration phase, for instance
+by calling SSSD helper process `fido2_child --register`.
```

Comment:
We would like to avoid calling the SSSD helper process directly, and instead call the `sssctl` wrapper. So I would change a little bit the wording and also mention that the call is `sssctl fido2 --register`

Reply:
`sssctl` currently is not usable for non-privileged users. It assumes it is run by root and fails to do so. Since the registration process happens at the client side, `ipa` CLI can and most likely be used by a non-privilged user. Thus, we cannot use `sssctl` directly.

Reply:
Reworded to mention sssctl